### PR TITLE
mia: improve CIC detection algorithm

### PR DESCRIPTION
**Pull request by rasky**
Currently, we detect the CIC to emulate by calculating a CRC32 over IPL3 bootcode in ROM, and keeping a database of the known IPL3s. This works because right now there are only a handful of IPL3 variants (exactly one per CIC).

In preparation for libdragon releasing an open source IPL3 that could evolve and be forked an unbounded number of times, we must stop keeping a database of IPL3s.

This commit changes strategy: it simulates the IPL2 checksum hash over IPL3 (which is exactly what a real N64 does at boot) and verifies if the checksum matches that expected by any of the known CIC variants. This works because open source IPL3 variants must anyway have a checksum that collides with that of CICs (which must be obtained via a second preimage attack, bruteforced on GPUs), otherwise they would not work on real hardware nor on Ares itself (which correctly simulates the boot sequence and would refuse to boot an "unsigned" IPL3).

### CODE REVIEW

File: mia/medium/nintendo-64.cpp [hunk 1]
```diff
@@ -4,8 +4,97 @@ struct Nintendo64 : Cartridge {
   auto load(string location) -> bool override;
   auto save(string location) -> bool override;
   auto analyze(vector<u8>& rom) -> string;
+  auto ipl2checksum(u32 seed, array_view<u8> rom) -> u64;
 };
 
+auto Nintendo64::ipl2checksum(u32 seed, array_view<u8> rom) -> u64
+{
+  auto rotl = [](u32 value, u32 shift) -> u32 {
+    return (value << shift) | (value >> (32 - shift));
```
Comment (invertego) [inline mia/medium/nintendo-64.cpp]: Shifting a u32 by 32 bits is UB. Same for rotr.

Might be good to move the rotate helpers to nall/bit.hpp. We can eventually just switch to the C++ standard library implementation when we move to C++20.

https://en.cppreference.com/w/cpp/numeric/rotl

Comment (rasky) [reply mia/medium/nintendo-64.cpp]: Done, but I don't want to move it to nall as I don't want to think now how to make that a template.

File: mia/medium/nintendo-64.cpp [hunk 2]
```diff
@@ -4,8 +4,97 @@ struct Nintendo64 : Cartridge {
...
+  u32 state[16];
+  for (auto &s : state) s = init;
+
+  u32 dataNext, dataLast;
+  u32 dataIndex = 0;
+  u32 loop = 0;
+  while (1) {
+      loop++;
+      dataLast = data;
+      data = read(dataIndex);
```
Comment (invertego) [inline mia/medium/nintendo-64.cpp]: FYI, array_view has a built-in helper for big endian reads.
```
      data = rom.readm(4);
      rom += 4;
```

File: mia/medium/nintendo-64.cpp [hunk 3]
```diff
@@ -4,8 +4,97 @@ struct Nintendo64 : Cartridge {
   auto load(string location) -> bool override;
   auto save(string location) -> bool override;
   auto analyze(vector<u8>& rom) -> string;
+  auto ipl2checksum(u32 seed, array_view<u8> rom) -> u64;
 };
 
+auto Nintendo64::ipl2checksum(u32 seed, array_view<u8> rom) -> u64
+{
```
Comment (invertego) [inline mia/medium/nintendo-64.cpp]: Nit: brace style

File: mia/medium/nintendo-64.cpp [hunk 4]
```diff
@@ -4,8 +4,97 @@ struct Nintendo64 : Cartridge {
...
+      state[10]  = csum(state[10] + data, dataNext, loop);
+      state[11]  = csum(state[11] ^ data, dataNext, loop);
+      state[12] += state[8] ^ data;
+  }
+
+  u32 buf[4];
+  for (auto &b : buf) b = state[0];
+
+  for (loop = 0; loop < 16; loop++)
+  {
```
Comment (invertego) [inline mia/medium/nintendo-64.cpp]: Nit: brace style

**Review by invertego — COMMENTED** [review]
Looks good, just some minor suggestions.

File: mia/medium/nintendo-64.cpp (hunk 2 above)
Comment (rasky) [reply mia/medium/nintendo-64.cpp]: Done, though `readm` already advances the pointer

File: mia/medium/nintendo-64.cpp (hunk 3 above)
Comment (rasky) [reply mia/medium/nintendo-64.cpp]: Done

File: mia/medium/nintendo-64.cpp (hunk 4 above)
Comment (rasky) [reply mia/medium/nintendo-64.cpp]: Done

**Pull request merged**